py/scheduler.c

```diff
@@ -99,6 +100,9 @@ static inline void mp_sched_run_pending(void) {
         MICROPY_END_ATOMIC_SECTION(atomic_state);
         callback(node);
         atomic_state = MICROPY_BEGIN_ATOMIC_SECTION();
+        if (node == original_tail) {
+            break; // Don't execute any callbacks scheduled during this run
+        }
```

Comment:
Can this be rewritten as:
```c
mp_sched_node_t *original_tail = MP_STATE_VM(sched_tail);
while (MP_STATE_VM(sched_head) != original_tail) {
    ...
}
```
?

Reply:
No, that way the scheduler always skips the last entry in the list. I think it could be written slightly differently as:

```c
mp_sched_node_t *original_tail = MP_STATE_VM(sched_tail);
if (MP_STATE_VM(sched_head) != NULL) {
    do {
      ..
  } while (MP_STATE_VM(sched_head) != original_tail);
}
```

(Provided the list entries aren't ever modified during the run - apart from appending - which is currently the case.)

This takes the comparison to NULL out of the loop body. Unsure if that'd make much difference in the binary, but I can test it out if you like this way.

Reply:
(Edited the above as I think can remove a pointer lookup by comparing `original_tail` to `NULL` instead of `MP_STATE_VM(sched_head)`.)

Reply:
In a local build, code size for rp2 was the same as the current version after I moved the `original_tail` assignment outside of the loop.